src/Instruction.cpp

```diff
@@ -360,113 +131,96 @@ void BitArray::recoverFromEndianness(unsigned alignment, endian_t endianness)
 }
 
-BitArrayRange::BitArrayRange(unsigned startindex_included, unsigned endindex_included)
-    : filterStart_(startindex_included), filterEnd_(endindex_included), filterLen_(filterStart_ + 1 - filterEnd_)
-{
-    assert(filterStart_ + 1 >= filterEnd_ && "Invalid BitArrayRange constructor arguments");
-    assert(filterLen_ <= sizeof(I) * 8 && "Invalid BitArrayRange width");
+void BitArray::set_value(size_type width, unsigned long value){
+    BitArray b(width, value);
+    *this = b;
+}
 
-    lowPartShift_ = filterEnd_ % Ibits;
-    if (lowPartShift_ + filterLen_ <= Ibits)
-    {
-        needsSplitAccess_ = false;
-        dataArrayIndex_ = (filterEnd_ - lowPartShift_) / Ibits;
-        for (unsigned i = 0; i < filterLen_; i++)
-        {
-            lowPartMask_ = (lowPartMask_ << 1) | 1;
-        }
-    }
-    else
-    {
-        needsSplitAccess_ = true;
-        highPartShift_ = Ibits - lowPartShift_;
-        dataArrayIndex_ = (filterEnd_ - lowPartShift_) / Ibits;
-        for (unsigned i = 0; i < highPartShift_; i++)
-        {
-            lowPartMask_ = (lowPartMask_ << 1) | 1;
-        }
-        for (unsigned i = 0; i < (lowPartShift_ + filterLen_) * Ibits; i++)
-        {
-            highPartMask_ = (highPartMask_ << 1) | 1;
-        }
-    }
+void BitArray::set_value(unsigned long value){
+    set_value(size(), value);
 }
 
-I BitArrayRange::read(const BitArray &ba)
+BitArray BitArray::get_range(size_type end, size_type start) const
 {
-    assert(ba.w_ > filterStart_ && "BitArrayRange outside of BitArray");
-
-    I ret = (ba.d_[dataArrayIndex_] >> lowPartShift_) & lowPartMask_;
-    if (needsSplitAccess_)
-    {
-        ret |= (ba.d_[dataArrayIndex_ + 1] & highPartMask_) << highPartShift_;
-    }
+    // shift the bits and resize it accordingly
+    BitArray ret = *this;
+    ret >>= start;
+    ret.resize(end-start+1);
     return ret;
 }
-void BitArrayRange::write(const BitArray &ba, I val)
+
+void BitArray::set_range(unsigned long val, size_type end, size_type start){
+    auto len = end - start + 1;
+    BitArray range(len, val);
+    for(size_type i = 0; i < len; ++i)
+        (*this)[i+start] = range[i];
+}
+
+std::string BitArray::to_string() const
 {
-    assert(ba.w_ > filterStart_ && "BitArrayRange outside of BitArray");
+    std::string s;
+    boost::to_string(*this, s);
+    return s;
+}
 
-    ba.d_[dataArrayIndex_] =
-        ((val & lowPartMask_) << lowPartShift_) | (ba.d_[dataArrayIndex_] & ~(lowPartMask_ << lowPartShift_));
-    if (needsSplitAccess_)
-    {
-        ba.d_[dataArrayIndex_ + 1] =
-            ((val >> highPartShift_) & highPartMask_) | (ba.d_[dataArrayIndex_ + 1] & ~highPartMask_);
-    }
+BitArrayRange::BitArrayRange(unsigned endindex_included, unsigned startindex_included)
+    : startpos(startindex_included), endpos(endindex_included) {}
+
+I BitArrayRange::read(const BitArray& ba)
+{
+    assert(ba.size() > (endpos - startpos + 1) && "BitArrayRange outside of BitArray");
```

Comment:
This assertion would be raised when accessing `X[31:0]` while X is 32-bits large.

Shouldn't the comparison be `>=` instead of `>`.

(Sorry that it took me over a year to find this...)